lib/chef/provisioning/aws_driver/driver.rb

```diff
@@ -170,23 +170,25 @@ def allocate_load_balancer(action_handler, lb_spec, lb_options, machine_specs)
         end
       end
 
-      # Update instance list
+      # Update instance list, but only if there are machines specified
       actual_instance_ids = Set.new(actual_elb.instances.map { |i| i.instance_id })
 
-      instances_to_add = machine_specs.select { |s| !actual_instance_ids.include?(s.location['instance_id']) }
-      instance_ids_to_remove = actual_instance_ids - machine_specs.map { |s| s.location['instance_id'] }
-
-      if instances_to_add.size > 0
-        perform_action.call("  add machines #{instances_to_add.map { |s| s.name }.join(', ')}") do
-          instance_ids_to_add = instances_to_add.map { |s| s.location['instance_id'] }
-          Chef::Log.debug("Adding instances #{instance_ids_to_add.join(', ')} to load balancer #{actual_elb.name} in region #{aws_config.region}")
-          actual_elb.instances.add(instance_ids_to_add)
+      if machine_specs
```

Comment:
I think what you've done caller side is turn "not specifying machines" into
"specifying NO machines" - i.e. equivalent of saying "machines([]) - right?

If so, the problem is that whenever the recipe runs, the load balancer will
try to detach the auto-scaling instances (triggering the AWS FormatError).
If we got past the error, then the aws-auto-scaling-group resource would
need to reattach the instances.

 So that's why I think we need a way to say "there needs to be a load
balancer, but dont do anything to its attached instances".

Regards,
Christine

On Mon, Mar 2, 2015 at 1:13 PM, John Keiser <notifications@github.com>
wrote:

> In lib/chef/provisioning/aws_driver/driver.rb
> <https://github.com/chef/chef-provisioning-aws/pull/117#discussion_r25627762>
> :
>
> >        actual_instance_ids = Set.new(actual_elb.instances.map { |i| i.instance_id })
> >
> > -      instances_to_add = machine_specs.select { |s| !actual_instance_ids.include?(s.location['instance_id']) }
> > -      instance_ids_to_remove = actual_instance_ids - machine_specs.map { |s| s.location['instance_id'] }
> > -
> > -      if instances_to_add.size > 0
> > -        perform_action.call("  add machines #{instances_to_add.map { |s| s.name }.join(', ')}") do
> > -          instance_ids_to_add = instances_to_add.map { |s| s.location['instance_id'] }
> > -          Chef::Log.debug("Adding instances #{instance_ids_to_add.join(', ')} to load balancer #{actual_elb.name} in region #{aws_config.region}")
> > -          actual_elb.instances.add(instance_ids_to_add)
> > +      if machine_specs
>
> Do you have the latest chef-provisioning? I ask because I think I fixed
> this issue on the caller side:
> https://github.com/chef/chef-provisioning/blob/master/lib/chef/provider/load_balancer.rb#L28
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/chef/chef-provisioning-aws/pull/117/files#r25627762>.
>